docs/understanding-airbyte/airbyte-protocol.md

```diff
@@ -0,0 +1,735 @@
+
+#### Input:
+1. `config` - A configuration JSON object that has been validated using `ConnectorSpecification#connectionSpecification` (see [ActorSpecification](#actor-specification) for information on `connectionSpecification`).
+
+#### Output:
+1. `connectionStatus` - an [AirbyteConnectionStatus Message](#airbyteconnectionstatus-message) wrapped in an `AirbyteMessage` of type `connection_status`.
+
+## Source
+
+A Source is an application that extracts data from an underlying data store. A Source implements the following interface:
+
+```
+spec() -> ConnectorSpecification
+check(Config) -> AirbyteConnectionStatus
+discover(Config) -> AirbyteCatalog
+read(Config, ConfiguredAirbyteCatalog, State) -> Stream<AirbyteRecordMessage | AirbyteStateMessage>
+```
+
+`spec` and `check` are the same as the commands described in the [Common Commands](#common-interface) section.
+
+### Discover
+The `discover` method detects and describes the _structure_ of the data in the data store and which Airbyte configurations can be applied to that data.
+
+#### Input:
+1. `config` - A configuration JSON object that has been validated using `ConnectorSpecification#connectionSpecification` (see [ActorSpecification](#actor-specification) for information on `connectionSpecification`).
+
+#### Output:
+1. `catalog` - an [AirbyteCatalog](#catalog) wrapped in an `AirbyteMessage` of type `catalog`. See the [Catalog Section](#catalog) for details.
+
+### Read
+`read` extracts data from the underlying data store and emits it as `AirbyteRecordMessage`s. It also emits `AirbyteStateMessage`s to allow checkpointing replication.
+
+#### Input:
+1. `config` - A configuration JSON object that has been validated using `ConnectorSpecification#connectionSpecification` (see [ActorSpecification](#actor-specification) for information on `connectionSpecification`).
+2. `configured catalog` - A `ConfiguredAirbyteCatalog` is built on top of the `catalog` returned by `discover`. The `ConfiguredAirbyteCatalog` specifies HOW the data in the catalog should be replicated. The catalog is documented in the [Catalog Section](#catalog).
+3. `state` - An JSON object that represents a checkpoint in the replication. This object is only ever written or read by the source, so it is a JSON blob with whatever information is necessary to keep track of how much of the data source has already been read (learn more in the [State & Checkpointing](#state--checkpointing) Section).
+
+#### Output:
+1. `message stream` - A stream of `AirbyteRecordMessage`s and `AirbyteStateMessage`s piped to stdout.
+   * A source outputs `AirbyteStateMessages` in order to allow checkpointing data replication. State is described in more detail below in the [State & Checkpointing](#state--checkpointing) section.
+   * Only `AirbyteRecordMessage`s that contain streams that are in the catalog will be processed. Those that do not will be ignored. See [Schema Mismatches](#schema-mismatches) for more details.
```

Comment:
... `AirbyteRecordMessage`s from multiple streams can be multiplexed/mixed together, and do not need to be emitted serially as a group.

Reply:
:+1: